src/Microsoft.AspNet.Mvc.Core/Routing/UrlHelper.cs

```diff
@@ -177,12 +162,16 @@ protected virtual string GeneratePathFromRoute(string routeName, RouteValueDicti
                     }
 
                     builder.Append(pathData.VirtualPath);
```

Comment:
Slightly unrelated to your change but the `VirtualPath` handling is doing about the same things in two completely different ways. Suggest making the `if` and `else` blocks more consistent though of course `VirtualPath.Length == 0` isn't quite the same w/ and w/o a `PathBase`.